inc/namespace.php

```diff
@@ -456,3 +459,92 @@ function on_request_stats( TransferStats $stats ) {
 
 	on_aws_guzzle_request_stats( $stats );
 }
+
+/**
+ * Get the EC2 Instance metadata.
+ *
+ * This will cause a remote request to the metadata service when the cache is empty.
+ *
+ * @return array $array(
+ *    accountId: string,
+ *    architecture: string,
+ *    availabilityZone: string,
+ *    billingProducts: string,
+ *    devpayProductCodes: string,
+ *    marketplaceProductCodes: string,
+ *    imageId: string,
+ *    instanceId: string,
+ *    instanceType: string,
+ *    kernelId: string,
+ *    pendingTime: string,
+ *    privateIp: string,
+ *    ramdiskId: string,
+ *    region: string,
+ *    version: string,
+ * )
+ */
+function get_ec2_instance_metadata() : array {
+	$has_cache = false;
+	$cache_key = 'altis.ec2_instance_metadata';
+	$cached_data = apcu_fetch( $cache_key, $has_cache );
+	if ( $has_cache ) {
+		return $cached_data;
+	}
+
+	$client = new Client();
+
+	try {
+		$request = $client->request( 'GET', 'http://169.254.169.254/latest/dynamic/instance-identity/document', [
+			'timeout' => 1,
+			'on_stats' => __NAMESPACE__ . '\\on_request_stats',
+		] );
+	} catch ( Exception $e ) {
+		trigger_error( sprintf( 'Unable to get instance metadata. Error: %s', $e->getMessage() ), E_USER_NOTICE );
+		apcu_store( $cache_key, [] );
+		return [];
+	}
+
+	if ( $request->getStatusCode() !== '200' ) {
+		trigger_error( sprintf( 'Unable to get instance metadata. Returned response code: %s', $request->getStatusCode() ), E_USER_NOTICE );
+		apcu_store( $cache_key, [] );
+		return [];
+	}
+
+	$metadata = json_decode( $request->getBody(), true );
+
+	if ( ! $metadata ) {
+		$metadata = [];
+	}
+
+	apcu_store( $cache_key, $metadata );
+
+	return $metadata;
+}
+
+/**
+ * Add the EC2 instance data to the Xray root segment.
+ *
+ * @param array $trace
+ * @return array
+ */
+function add_ec2_instance_data_to_xray( array $trace ) : array {
+	// Only add instance information to the root trace.
+	if ( ! empty( $trace['parent_id'] ) ) {
+		return $trace;
+	}
+
+	$metadata = get_ec2_instance_metadata();
+	if ( ! $metadata ) {
+		return $trace;
+	}
+
+	$trace['aws']['ec2']['availability_zone'] = $metadata['availabilityZone'];
```

Comment:
It's a bit confusing that we can get both ec2 and ecs instance metadata, I thought they were somewhat exclusive from an infrastructure standpoint. Can you add a few comments to note why there's ec2 data when using ecs / where it comes from? Will there be ec2 metadata when using ecs?

Reply:
No, ecs runs on ec2, so in the case of running on ecs, we want to send both the underlaying instance data, and the ecs container id.

Reply:
Ah cool, makes sense of course. I thought it was maybe a bit more abstracted than that 👍